### ISSUE TITLE
Chaining pattern when using setters

### ISSUE THREAD

**Issue by tasn** [Dushistov/flapigen-rs#329]
Hi I'm trying to figure out how to implement chaining of setters. Currently what I have in Rust is:
```
pub struct Foobar {
    a: usize,
    b: bool,
}

impl Foobar {
    pub fn new() -> Self {
        Self {
            a = 0,
            b = false,
        }
    }

    pub fn set_a(mut self, a: usize) -> Self {
        self.a = a;
        self
    }

    pub fn set_b(mut self, b: bool) -> Self {
        self.b = b
        self
    }
}
```

Which lets me init the sturct in a nice way, e.g:
```
let foobar = Foobar::new().set_a(23).set_b(true);
```
I want to have the same functionality in Java but I can't figure out how to do it. `self as the parameter` isn't supported and neither is `&mut self`

Thanks for your help, and thanks again for this great project!

**Comment by Dushistov** [comment]
self/mut self without '&' is not supported, and I doubt that it would. Java/C++ is not supported methods that can consume object,
may be it is possible with Python, not sure. If map such methods to static function that obviously semantic would be broken.

May be return of `&mut Self` will be supported for C++ case (#318), but I am not sure about Java.
Returning references in Java is not the great thing, garbage collection can come at any moment.
And there is no guarantee that user of Java code does not do strange thing like:

```Java
Foobar f = new Foobar();
Foobar f2 = f.set_a(23);
```
and then use `f2` instead of `f`, and `f2` becomes invalid after GC free memory for `f`.

**Comment by tasn** [comment]
Why wouldn't your example work with &mut? Java won't be giving out it's reference (just a mut borrow) so f will retain the object and so will f2. They will both point to the same object. The only thing we need to do to handle this is refcounting behind the scenes.

**Comment by Dushistov** [comment]
> They will both point to the same object
> this is refcounting behind the scenes

For obvious reasons, I prefer explicitly indicate the possible overhead.
So if user need refcounting on Rust side, he/she should instruct flapigen use
`Arc` or `Rc` for holding the object instead of default `Box`.
he/she also have to write bunch of inline methods for pattern chaining:

```rust
foreign_class!(class Foobar {
    self_type Foobar;
    constructor new() -> Arc<Mutex<Foobar>> {
        Arc::new(Mutex::new(Foobar::new()))
    }
    fn setA(self: Arc<Mutex<Foobar>>, a: usize) -> Arc<Mutex<Foobar>> {
        todo!()
    } 
``` 

Note: `self: Arc<>` is not supported in the current version, it is just idea.

But even if flapigen would support such kind of syntax,
there are still problem you can not move out of Arc call `set_a` and then move object back,
you need `&mut self` and `&mut Self`.

If you want use refcounting of the Java side, you can do it with current version of flapigen:

```rust
foreign_class!(class Foobar {
    self_type Foobar;

    constructor Foobar::new() -> Foobar;
    fn do_setA(&mut self, a: usize) {
        this.set_a(a);
    }
    fn do_setB(&mut self, b: bool) {
        this.set_b(b);
    }
    foreign_code r#"
    public Foobar setA(long a) {
        this.do_setA(a);
        return this;
    }
"#;
});
```

**Comment by tasn** [comment]
I agree about explicitly indicating possible overhead *in rust*, though I think this is not relevant when you have bindings. The cost of traversing the JNI/CPython bridge are orders of magnitude higher than a refcount.

Anyhow, I wasn't talking about solving it in the rust side, I don't think it's possible (is it?). It needs to be solved in the flapigen side, and just the generated code should generate it in the same way it would have been generated natively.

So e.g. when you detect a function returning `&mut self`, automatically understand it as: keep a refcount of the rust struct inside of the java object. This way there will be no copying (just ref up+down) when returning such values, and the Java side would be nice too. The builder pattern is a very common one in Java, and not having it makes the code quite ugly and cumbersome to use.

**Comment by tasn** [comment]
Here is what I ended up implementing manually (in Python), though there's no reason why flapigen couldn't do it automatically:
https://github.com/etesync/etebase-py/blob/319939e109d14dc22b22132dd7a50e4dfde36de8/etebase/__init__.py#L182
